File: src/main/java/edu/cornell/cs/apl/viaduct/util/MeetSemiLattice.java
package edu.cornell.cs.apl.viaduct.util;

/** A set that supports binary greatest lower bounds. */
public interface MeetSemiLattice<T extends MeetSemiLattice<T>> extends PartialOrder<T> {
  /** Greatest lower bound of {@code this} and {@code that}. */
  T meet(T that);
}

File: compiler/src/main/kotlin/io/github/apl_cornell/viaduct/algebra/MeetSemiLattice.kt
package io.github.apl_cornell.viaduct.algebra

/** A set that supports binary greatest lower bounds. */
interface MeetSemiLattice<T : MeetSemiLattice<T>> {
    /** The greatest lower bound of `this` and [that]. */
    infix fun meet(that: T): T
}